### ISSUE TITLE
How could we approach mocking of dependencies?

### ISSUE THREAD

**Issue by retailcoder** [rubberduck-vba/Rubberduck#1550]
So we have nice boilerplate-free test methods, the ability to run setup and teardown methods for both every test method and every test module in a test suite, we'll have _data-driven tests_ by 2.1 (#1229), and 2.0 introduces new `@IgnoreTest` annotations and the ability to ignore specific tests without removing them.

The API exposes `AssertClass` and `PermissiveAssertClass` implementations of the `IAssert` interface; let's discuss how could we go about a _mocking_ API.

**Comment by MDoerner** [comment]
One possible step towards mocking dependencies could be to have a code generator for mock classes using the signature of a class to generate a corresponding mock class, which then provides dummy return values and saves how often and with which input values the class's methods have been called. 

I was thinking of something like the following:

Given a class like Class1 below
```vba
Option Explicit


Public Function Foo1(Input1 As SomeObjectType1, Input2 As SomeNonObjectType1) As SomeObjectType2
End Function


Public Function Foo2() As SomeNonObjectType2
End Function


Public Sub Soo(Input1 As SomeObjectType3)
End Sub
```
it might generate a class MockClass1 somewhat like 
```vba
Option Explicit

Implements Class1


Private Const strKeyPartSeparator As String = "_akjd_"  'Something random hopefully nobody uses in paramater names or method names

Private dicReturnValues As Dictionary
Private dicInputValues As Dictionary
Private dicCallCounts As Dictionary
Private dicInputValueIsObjectType As Dictionary
Private dicReturnValueIsObjectType As Dictionary



'Constructor

Private Sub Class_Initialize()
    InitializeReturnValuesDictionary
    InitializeCallCountDictionary
    InitializeInputValuesDictionary
    InitializeInputValueTypeClassDictionary
    InitializeReturnValueTypeClassDictionary
End Sub

    Private Sub InitializeReturnValuesDictionary()
        Set dicReturnValues = New Dictionary
        SetDefaultReturnValues
    End Sub

        Private Sub SetDefaultReturnValues()   'TODO: Verify default mock return values.
            SetReturnValue "Foo1", Nothing
            LetReturnValue "Foo2", SomethingSensibleForTheSpecificType
        End Sub


    Private Sub InitializeCallCountDictionary()
        Set dicCallCounts = New Dictionary
        dicCallCounts("Foo1") = 0
        dicCallCounts("Foo2") = 0
        dicCallCounts("Soo") = 0
    End Sub
     

    Private Sub InitializeInputValuesDictionary()
        Set dicInputValues = New Dictionary
    End Sub
    
    
    Private Sub InitializeInputValueTypeClassDictionary()
        Set dicInputValueIsObjectType = New Dictionary
        dicInputValueIsObjectType(InputParameterKey("Foo1", "Input1")) = True
        dicInputValueIsObjectType(InputParameterKey("Foo1", "Input2")) = False
        dicInputValueIsObjectType(InputParameterKey("Soo", "Input1")) = True
    End Sub
    
    
    Private Sub InitializeReturnValueTypeClassDictionary()
        Set dicReturnValueIsObjectType = New Dictionary
        dicReturnValueIsObjectType("Foo1") = True
        dicReturnValueIsObjectType("Foo2") = False
    End Sub




'Public Methods

Public Sub SetReturnValue(ByVal MethodName As String, ByVal Value As Variant)
    Set dicReturnValues(MethodName) = Value
End Sub


Public Sub LetReturnValue(ByVal MethodName As String, ByVal Value As Variant)
    dicReturnValues(MethodName) = Value
End Sub


Public Property Get ReturnValue(ByVal MethodName As String) As Variant
    If HasObjectTypeReturnValue(MethodName) Then
        Set ReturnValue = dicReturnValues(MethodName)
    Else
        ReturnValue = dicReturnValues(MethodName)
    End If
End Property



Public Property Get CallCount(ByVal MethodName As String) As Long
    CallCount = dicCallCounts(MethodName)
End Property



Public Property Get InputValues(ByVal MethodName As String, ByVal InputParameterName As String, ByVal ExecutionCount As Long) As Variant
    If IsObjectTypeInputParameter(MethodName, InputParameterName) Then
        Set InputValues = dicInputValues(InputValueInstanceKey(MethodName, InputParameterName, ExecutionCount))
    Else
        InputValues = dicInputValues(InputValueInstanceKey(MethodName, InputParameterName, ExecutionCount))
    End If
End Property




'Private Methods

Private Function IsObjectTypeInputParameter(ByVal MethodName As String, ByVal InputParameterName As String) As Boolean
    IsObjectTypeInputParameter = dicInputValueIsObjectType(InputParameterKey(MethodName, InputParameterName))
End Function



Private Function HasObjectTypeReturnValue(ByVal MethodName As String) As Boolean
    HasObjectTypeReturnValue = dicReturnValueIsObjectType(MethodName)
End Function



Private Function InputValueInstanceKey(ByVal MethodName As String, ByVal InputParameterName As String, ByVal ExecutionCount As Long) As String
   InputValueInstanceKey = InputParameterKey(MethodName, InputParameterName) & strKeyPartSeparator & ExecutionCount
End Function

    Private Function InputParameterKey(ByVal MethodName As String, ByVal InputParameterName As String) As String
        InputParameterKey = MethodName & strKeyPartSeparator & InputParameterName
    End Function



Private Sub IncrementCallCount(ByVal MethodName As String)
    dicCallCounts(MethodName) = dicCallCounts(MethodName) + 1
End Sub



Private Sub SetOrLetInputValue(ByVal MethodName As String, ByVal InputParameterName As String, ByVal CurrentCallCount As Long, ByVal Value As Variant)
    If IsObjectTypeInputParameter(MethodName, InputParameterName) Then
        SetInputValue MethodName, InputParameterName, CurrentCallCount, Value
    Else
        LetInputValue MethodName, InputParameterName, CurrentCallCount, Value
    End If
End Sub

    Private Sub SetInputValue(ByVal MethodName As String, ByVal InputParameterName As String, ByVal CurrentCallCount As Long, ByVal Value As Variant)
        Dim strKey As String
        strKey = InputValueInstanceKey(MethodName, InputParameterName, CurrentCallCount)
        Set dicInputValues(strKey) = Value
    End Sub
     
    
    Private Sub LetInputValue(ByVal MethodName As String, ByVal InputParameterName As String, ByVal CurrentCallCount As Long, ByVal Value As Variant)
        Dim strKey As String
        strKey = InputValueInstanceKey(MethodName, InputParameterName, CurrentCallCount)
        dicInputValues(strKey) = Value
    End Sub



Private Sub AssignReturnValue(ByRef ReturnValueVariable As Variant, ByVal MethodName As String)
    If HasObjectTypeReturnValue(MethodName) Then
        Set ReturnValueVariable = ReturnValue(MethodName)
    Else
        ReturnValueVariable = ReturnValue(MethodName)
    End If
End Sub



'Mocked Methods

Private Function Class1_Foo1(Input1 As SomeObjectType1, Input2 As SomeNonObjectType1) As SomeObjectType2
    Dim strMethodName As String
    strMethodName = "Foo1"
    IncrementCallCount strMethodName
    SetOrLetInputValue strMethodName, "Input1", CallCount(strMethodName), Input1
    SetOrLetInputValue strMethodName, "Input2", CallCount(strMethodName), Input2
    AssignReturnValue Class1_Foo1, strMethodName
End Function



Private Function Class1_Foo2() As SomeNonObjectType2
    Dim strMethodName As String
    strMethodName = "Foo2"
    IncrementCallCount strMethodName
    AssignReturnValue Class1_Foo2, strMethodName
End Function



Private Sub Class1_Soo(Input1 As SomeObjectType3)
    Dim strMethodName As String
    strMethodName = "Soo"
    IncrementCallCount strMethodName
    SetOrLetInputValue strMethodName, "Input1", CallCount(strMethodName), Input1
End Sub
```
Obviously, this only works for classes whose interface can be implemented properly.

**Comment by MDoerner** [comment]
In my suggestion above, the public members should probably go into some interface, `IMock` say. Moreover, the the storage handling should probably be refactored into its own class behind an interface, `IMockBase` say. Then the generated code would only need to contain the class specific setup and delegations to the base mock.

**Comment by Vogel612** [comment]
The usual approach for Mocks in Java (as a perspective) is extending the mocked class and overriding the methods that are mocked with the given behaviour. All interactions with those classes are "logged" and later verified.

This usually happens with Runtime code rewriting over reflection (let's call it black magic). It might be possible to simplify a bit with Lambdas ... Either way, Mocking frameworks are usually just libraries. They aren't generally integrated in the IDE, nor are they in a separate language. 

Since Unit-Testing frameworks are also written in Java and just imported and then run with black magic (annotations this time) it's unreasonable to expect us to perform the same functionality with VBA. 
If we only need to mock Reference Objects, we should IMO do it on the other side of the COM boundary (read in C#). 
Overall this is a nice feature, but it's going to be quite the black magic we have there. I'd prefer to keep this (and possibly Unit-Testing) separate from the add-in itself. Yes I'm trying to push the plugin architecture here :smile_cat:

**Comment by retailcoder** [comment]
@Vogel612 I agree, at the end of the day this would end up being exposed as a COM API, and I fully support the plugin architecture idea.

The difficulty with mocking VBA interfaces is that, well, there's no class inheritance in VBA, so we can't extend and override the way Java or C# does.

If we do the mocks in C#, we'll need them COM-visible - and COM interop doesn't support generics, so we'd need to late-bind everything:

    Dim mockSomething As Object
    Set mockSomething = Rubbermock.Create("VBAProject.Something")

And then on the managed side, black magic:

 - Use the supplied string to get ahold of the specified class module
 - Use Rubberduck's parser to get ahold of the class' public interface
 - Use reflection (C#) to emit a corresponding COM-visible interface
 - Emit an object that implements the COM-visible interface
 - `Setup` make the implementation increment some member call counter when VBA code calls it
 - `Returns` make the implementation return a specified constant value/object

The way I'm seeing it, the base class would expose the emitted object a bit like Moq does:

    Dim mockSomething As Object
    Set mockSomething = Rubbermock.Create("VBAProject.SomeInterface")
    mockSomething.Setup("GetFoo", Rubberduck.AnyInteger).Returns(12)
    mockSomething.Setup("DoBar")

    Dim sut As MyClassUnderTest
    sut.DoSomething(mockSomething.Object)

    ' any way to avoid the assert call here without rewriting the whole unit testing feature?
    Assert.IsTrue mockSomething.Verify("DoBar")

I can't help but feel like this is a big, big can of worms right now, and designing the API so that it's both easy to use in VBA *and* actually useful - how could we do something like this?

    Dim mockMsgBox As Rubberduck.Rubbermock
    Set mockMsgBox = Rubbermock.Create("IMsgBox")
    mockMsgBox.Setup("Show", Rubberduck.AnyString, Rubberduck.AnyInteger, Rubberduck.AnyString)
              .Returns(vbYes)
    
    Dim sut As New SomeClassUnderTest
    ' any Yes/No IMsgBox with a prompt and a title should just say vbYes without prompting now:
    sut.DoSomething(mockMsgBox.Object)

More specifically, the chained nature of the setup calls makes me wonder: say `.Setup` returns some `IMockSetup` object, which exposes the `.Returns` method - I'm thinking of some kind of builder pattern, but is it right that an `IMockSetup` implementation alters the state of emitted code?

When should the code be emitted exactly? Immediately in the `Create` call, and then manipulated with every new `Setup` call? Or delayed until `.Object` is retrieved, so we know everything we need to know about the code to emit at the time we need to emit the mocked interface?

Do we emit C# or VBA code? If we emit C# code, how can we implement a VBA class/interface in C# (e.g. `IMsgBox` above)? If we emit VBA code, then we have the problem of needing to create an instance of a COM type from C# - if that was easy, unit testing would work off class modules already.

Feels like a dead end.

**Comment by retailcoder** [comment]
@MDoerner that's one approach; what I dislike is that it creates a bunch of actual VBA class modules in the user's VBA project, with the potential of pretty much doubling the number of code files in the project. Of course with `'@Folder` annotations we could dump 'em all into `'@Folder "Tests.Mocks"` and call it a day (it *would* indeed look fine in the RD Code Explorer), but then when opened without Rubberduck the sheer number of modules in the project could give a novice a heart attack!

At the same time I like that the mocks are readily usable, I'm not sure I like them statically defined and maintained; I'd much prefer have test setup code instruct RD how to setup the mocking.

As for what's "mockable", without a doubt it has to only be user interfaces. And since in VBA every class is an interface, that would mean any public member of any class could be "set-up"; I'd limit it to user code though, because I don't think we want to get into generating code for setting up the return value of `Worksheet.Range` calls (although, I'd question the need for a mock worksheet in the first place in that case; use a fake/stub instead, and run the test against a "disposable" worksheet that gets created for the test and then gets destroyed).

**Comment by MDoerner** [comment]
@Vogel612 I agree that a solution on the other side of the COM boundary would be nice. I just think that solving this on the VBA side might be a more pragmatic aproach.

@retailcoder I agree that having the possibility to instruct RD how to set up the mocking in the test itself would be far more elegant. Moreover, it would avoid the problem of cluttering projects with mock classes.

By the way, half of the setup in my example above is actually not needed in the end; I had not recognized that there is a native `IsObject` function in VBA. Using this function would also make the mock work properly with `Variant` return values and input parameters, of which one does not really know beforehand whether an object is provided.

As for what should be "mockable", it is probably anyway necessary to limit oneself to user interfaces as quite a number of the standard library objects cannot be implemented; you will get a compile time error trying to compile anything that implements their interface. (Usually, this is caused by some hidden members that cannot be implemented, e.g. because their name contains an underscore.)    
In my opinion, if someone really wants to mock some library class, he should build a simple adapter for it implementing a suitable user interface and mock that user interface.

**Comment by retailcoder** [comment]
> In my opinion, if someone really wants to mock some library class, he should build a simple adapter for it implementing a suitable user interface and mock that user interface.

Exactly. And I think RD could bundle up a number of such "adapter interfaces", e.g. `IMsgBox`, and perhaps even include one or two implementations - one for production code, another for test code.

Perhaps the "VBA fakes" approach is the most reasonable one here.

**Comment by Xipooo** [comment]
Not sure if the following is possible, but from a user perspective this would be nice.

Since VBA does have interfaces, make an interface a requirement of the Mocking.  Then annotate the interface with something like '@MockInterface allowing the RubberDuck parsing to detect the interface.   Also annotate each member with @MockMethod and @MockPropertySet/@MockPropertyGet.

```
'@MockInterface
Option Compare Database
Option Explicit

'@MockPropertyGet
Public Property Get MyProperty() as string
End Property

'@MockMethod
Public Function MyFunction() as boolean
End Function
```

Since an interface in VBA can be instantiated we could pass an instance of that class to a RubberDuck COM object in a function that would be called in the module initialize sub.  But since there isn't any logic behind the properties, the mock object would dictate the behavior.

```
'@TestModule
Private mock as new RubberDuck.MockClass
private myInterface as new IMyInterface

'@TestMethod
Public Sub TestMethod1()
     Dim SUT as new SystemUnderTest
     Dim result as variant

     //Arrange
     mock.SetInterface myInterface

     //Act
     result = SUT.DoSomething(mock.Object)

     //Assert
     mock.WasCalled "MyFunction"
End Sub
```

**Comment by retailcoder** [comment]
@Xipooo I think the only required annotation would be `@Interface` (already used to tell RD a class is an interface, when nothing implements it - when any class `Implements` it RD knows to treat it as an interface). What I'd love to do would be to `Setup` the mock, e.g.:

    mock.Setup("GetFoo").Returns 42

The thing is that we're not hooked into the runtime, so we can't inject a VBA type from C#..it's like there's a wall between the VBA execution context and RD - being able to collect the test results is a hack all by itself!

Basically the challenge is to make mock.Object implement the required interface.
